src/web/mjs/connectors/TenshiMoe.mjs

```diff
@@ -0,0 +1,154 @@
+        });
+        `;
+        await Engine.Request.fetchUI(request, script, 300000, true); //need image support, DG script loads a fake image
+    }
+
+    async _getMangaFromURI(uri) {
+        const request = new Request(uri, this.requestOptions);
+        const data = await this.fetchDOM(request, 'article.anime-entry header.entry-header h1');
+        const id = uri.pathname;
+        const title = data[0].textContent.trim();
+        return new Manga(this, id, title);
+    }
+
+    async _getMangas() {
+        let mangaList = [];
+        const uri = new URL('/anime', this.url);
+        const request = new Request(uri, this.requestOptions);
+        const data = await this.fetchDOM(request, 'ul.pagination li:nth-last-of-type(2) a');
+        const pageCount = parseInt(data[0].text);
+        for(let page = 1; page <= pageCount; page++) {
+            const mangas = await this._getMangasFromPage(page);
+            mangaList.push(...mangas);
+        }
+        return mangaList;
+    }
+
+    async _getMangasFromPage(page) {
+        const uri = new URL('/anime?page=' + page, this.url);
+        const request = new Request(uri, this.requestOptions);
+        const data = await this.fetchDOM(request, 'ul.loop.anime-loop li a');
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element, this.url),
+                title: element.title.trim()
+            };
+        });
+    }
+
+    async _getChapters(manga) {
+        let chapterlist = [];
+        const request = new Request(new URL(manga.id, this.url), this.requestOptions);
+        const data = await this.fetchDOM(request, 'ul.pagination li:nth-last-of-type(2) a');
+        const pageCount = data.length > 0 ? parseInt(data[0].text): 1;
+        for(let page = 1; page <= pageCount; page++) {
+            const chapters = await this._getchaptersFromPage(manga, page);
+            chapterlist.push(...chapters);
+        }
+        return chapterlist.reverse();
+    }
+
+    async _getchaptersFromPage(manga, page) {
+        const uri = new URL(manga.id +'?page='+ page, this.url);
+        const request = new Request(uri, this.requestOptions);
+        const data = await this.fetchDOM(request, 'ul.loop.episode-loop li a');
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element, this.url),
+                title: element.querySelector('div.episode-number').textContent.trim()
+            };
+        });
+    }
+
+    async _getPages(chapter) {
+        let request = new Request(new URL(chapter.id, this.url), this.requestOptions);
+        let data = await this.fetchDOM(request, 'ul.dropdown-menu a');
+        let sourceUrl = '';
+        if (data.length == 1) {
+            sourceUrl = data;
+        } else {
+            sourceUrl = this.findPreferedLanguage(data);
+            if(!sourceUrl) sourceUrl = data[0];
```

Comment:
Considere ternary operator `sourceUrl = sourceUrl ? sourceUrl : data[0];`
Doesn't `findPreferedLanguage` already return `data[0]` as default anyway?

Reply:
Yeah, turn out i am extra cautious for nothing.